node/integration_test.go

```diff
@@ -156,18 +174,74 @@ func TestTxGossipingAndAggregation(t *testing.T) {
 		assert.GreaterOrEqual(endCnt, adjustedHeight)
 		assert.GreaterOrEqual(commitCnt, adjustedHeight)
 
+		// GenerateFraudProof should have been called on each call to
+		// BeginBlock, DeliverTx, and EndBlock so the sum of their counts
+		// should be equal.
+		// Note: The value of clientNodes represents number of calls to DeliverTx
+		assert.Equal(beginCnt+clientNodes+endCnt, generateFraudProofCnt)
+
 		// assert that all blocks known to node are same as produced by aggregator
 		for h := uint64(1); h <= nodes[i].Store.Height(); h++ {
-			aggBlock, err := nodes[0].Store.LoadBlock(h)
-			require.NoError(err)
 			nodeBlock, err := nodes[i].Store.LoadBlock(h)
 			require.NoError(err)
+			aggBlock, err := nodes[0].Store.LoadBlock(h)
+			require.NoError(err)
+			// Only Intermediate state roots set by block aggregator are relevant, removed for sake of comparison
+			nodeBlock.Data.IntermediateStateRoots.RawRootsList = nil
 			assert.Equal(aggBlock, nodeBlock)
 		}
 	}
 }
 
-func createNodes(aggCtx, ctx context.Context, num int, wg *sync.WaitGroup, t *testing.T) ([]*Node, []*mocks.Application) {
+// Creates a starts the given number of client nodes along with an aggregator node. Uses the given flag to decide whether to have the aggregator produce malicious blocks.
+func createAndStartNodes(clientNodes int, isMalicious bool, t *testing.T) ([]*Node, []*mocks.Application) {
+	var wg sync.WaitGroup
+	aggCtx, aggCancel := context.WithCancel(context.Background())
+	ctx, cancel := context.WithCancel(context.Background())
+	nodes, apps := createNodes(aggCtx, ctx, clientNodes+1, true, &wg, t)
+	startNodes(nodes, &wg, t)
+	aggCancel()
+	time.Sleep(100 * time.Millisecond)
```

Comment:
[non-blocking]

As a follow up we should understand these sleeps more. Bare sleeps in testing usually lead to flaky tests in the future and people end up just bumping the sleep duration. 

Ideally we find a solution that is more of a retry function of checking for an expected state with a decently long timeout that we should only ever hit if there is actually something wrong.

Reply:
Same goes for all sleeps in these start up and create helpers

Reply:
Follow up: https://github.com/celestiaorg/rollmint/issues/569